lib/PlatoonService/src/V2VCommManager.cpp

```diff
@@ -0,0 +1,534 @@
+    }
+    else
+    {
+        /* Set topics. */
+        m_platoonHeartbeatTopic  = platoonHeartbeatTopicBuffer;
+        m_heartbeatResponseTopic = vehicleHeartbeatTopicBuffer;
+
+        LOG_DEBUG("Platoon Heartbeat Topic: %s", m_platoonHeartbeatTopic.c_str());
+        LOG_DEBUG("Vehicle Heartbeat Topic: %s", m_heartbeatResponseTopic.c_str());
+
+        if ((true == m_platoonHeartbeatTopic.isEmpty()) || (true == m_heartbeatResponseTopic.isEmpty()))
+        {
+            LOG_ERROR("Failed to create Platoon Heartbeat MQTT topics.");
+        }
+        else
+        {
+            /* Create lambda callback function for the platoon heartbeat topic. */
+            IMqttClient::TopicCallback lambdaHeartbeatInputTopicCallback = [this](const String& payload)
+            { this->platoonHeartbeatTopicCallback(payload); };
+
+            /* Subscribe to platoon heartbeat Topic. */
+            if (false == m_mqttClient.subscribe(m_platoonHeartbeatTopic, false, lambdaHeartbeatInputTopicCallback))
+            {
+                LOG_ERROR("Could not subcribe to MQTT Topic: %s.", m_platoonHeartbeatTopic.c_str());
+            }
+            else
+            {
+                isSuccessful = true;
+            }
+        }
+    }
+
+    return isSuccessful;
+}
+
+bool V2VCommManager::setupLeaderTopics()
+{
+    bool isSuccessful = false;
+
+    /* Participant is the leader. */
+    IMqttClient::TopicCallback lambdaVehiclesHeartbeatTopicCallback = [this](const String& payload)
+    { this->vehicleHeartbeatTopicCallback(payload); };
+
+    if (true == m_heartbeatResponseTopic.isEmpty())
+    {
+        LOG_ERROR("Failed to create Leader Heartbeat Response MQTT topic.");
+    }
+    /* Subscribe to Vehicles Heartbeat Topics. */
+    else if (false == m_mqttClient.subscribe(m_heartbeatResponseTopic, false, lambdaVehiclesHeartbeatTopicCallback))
+    {
+        LOG_ERROR("Could not subcribe to MQTT Topic: %s.", m_platoonHeartbeatTopic.c_str());
+    }
+    else
+    {
+        isSuccessful = true;
+    }
+
+    return isSuccessful;
+}
+
+bool V2VCommManager::sendPlatoonHeartbeat()
+{
+    bool isSuccessful = false;
+
+    /* Send platoon heartbeat. */
+    StaticJsonDocument<JSON_HEARTBEAT_MAX_SIZE> heartbeatDoc;
+    String                                      heartbeatPayload;
+    uint32_t                                    timestamp = millis();
+
+    heartbeatDoc["timestamp"] = timestamp;
+
+    if (0U == serializeJson(heartbeatDoc, heartbeatPayload))
+    {
+        LOG_ERROR("Failed to serialize heartbeat.");
+    }
+    else if (false == m_mqttClient.publish(m_platoonHeartbeatTopic, false, heartbeatPayload))
+    {
+        LOG_ERROR("Failed to publish MQTT message to %s.", m_platoonHeartbeatTopic.c_str());
+    }
+    else
+    {
+        LOG_DEBUG("Sent platoon heartbeat: %s", heartbeatPayload.c_str());
+        isSuccessful                    = true;
+        m_lastPlatoonHeartbeatTimestamp = timestamp;

```

Comment:
Comment whats the idea behind sending the timestamp.

Reply:
Done.